### PR TITLE
Backend: Added create room capability for authenticated users.

### PR DESCRIPTION
**Pull request by KH-Amir-TN**
## Description

Added create room capability for authenticated users.
## Testing Steps

1. Pull the code.
2. Install the dependencies `bundle install && npm|yarn install`.
3. Clean the previous assets build by running `rm app/assets/builds/*` (This won't remove .keep since it's hidden).
4. Clean the database and tmp files for a better isolation by running `rails tmp:clear && rails db:schema:cache:clear && rails db:migrate:reset`
5. Run the linter and specs `bundle exec rubocop --parallel && bundle exec rspec`
6. Run `./bin/dev` to run the assets builders processes and the Puma server all at once.

### CODE REVIEW

File: app/controllers/api/v1/rooms_controller.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@ module V1
     class RoomsController < ApplicationController
       skip_before_action :verify_authenticity_token # TODO: amir - Revisit this.
       before_action :find_room, only: %i[show start]
+      before_action :ensure_authenticated, only: :create
```
Comment (farhatahmad) [inline app/controllers/api/v1/rooms_controller.rb:8]: I would leave this out for now - we should have a separate PR implementing this correctly across all controllers

File: app/controllers/api/v1/rooms_controller.rb [hunk 2]
```diff
@@ -44,11 +44,24 @@ def start
         end
       end
 
+      # POST /api/v1/rooms.json
+      # Returns: { data: Array[serializable objects] , errors: Array[String] }
+      # Does: creates a room for the authenticated user.
+      def create
+        room = Room.new(room_params.merge(user_id: current_user.id)) and room.save!
```
Comment (farhatahmad) [inline app/controllers/api/v1/rooms_controller.rb:51]: This syntax is weird. Remember that keeping things simple makes life much easier for us

```
Room.create!(room_params.merge(user_id: current_user.id))
```

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -42,6 +42,10 @@ def render_json(data: {}, errors: [], status: :ok)
     }, status:
   end
 
+  def ensure_authenticated
+    render_json status: :unauthorized and return unless current_user
```
Comment (farhatahmad) [inline app/controllers/application_controller.rb:46]: Lets leave this for another PR (and avoid the `and` unless necessary)

File: spec/controllers/rooms_controller_spec.rb [hunk 4]
```diff
@@ -59,9 +59,29 @@
...
+  describe '#create' do
+    let(:room_params) do
+      {
+        room: { name: Faker::Science.science }
+      }
+    end
+
+    it 'creates a room for the authenticated user' do
+      session[:user_id] = user.id
+      expect { post :create, params: room_params }.to change { user.reload.rooms.count }.from(0).to(1)
```
Comment (farhatahmad) [inline spec/controllers/rooms_controller_spec.rb:77]: Is the `reload` necessary here?

File: app/controllers/api/v1/rooms_controller.rb (hunk 2 above)
Comment (KH-Amir-TN) [reply app/controllers/api/v1/rooms_controller.rb:51]: You got reason, 
It's funny though that I used it at first but don't why I've changed it.

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (KH-Amir-TN) [reply app/controllers/application_controller.rb:46]: Sure,
Does the `render` method kills the request/response chain, if so then the `and return` is absurd unlike with the `redirect_to`.

File: spec/controllers/rooms_controller_spec.rb (hunk 4 above)
Comment (KH-Amir-TN) [reply spec/controllers/rooms_controller_spec.rb:77]: After checking, It appears that it is not required.
The `count` method seems to be making its own `SQL query`.

File: app/controllers/api/v1/rooms_controller.rb (hunk 1 above)
Comment (KH-Amir-TN) [reply app/controllers/api/v1/rooms_controller.rb:8]: You want me to remove it?

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (KH-Amir-TN) [reply app/controllers/application_controller.rb:46]: Render doesn't stop the request/response chain, to avoid a full `ìf...else....end` block while using the one liner syntax using the `and return` to kill the chain is recommended.
https://guides.rubyonrails.org/layouts_and_rendering.html#avoiding-double-render-errors

Comment (farhatahmad) [reply app/controllers/application_controller.rb]:
```
def ensure_authenticated
  return unless current_user
  return render_json status: :unauthorized  
end
```

File: app/controllers/api/v1/rooms_controller.rb (hunk 1 above)
Comment (farhatahmad) [inline app/controllers/api/v1/rooms_controller.rb:8]: Don't forget to remove this

Comment (KH-Amir-TN) [reply app/controllers/api/v1/rooms_controller.rb:8]: Do you want me to call `#ensure_authenticated` manually in the `#create`?

Comment (KH-Amir-TN) [reply app/controllers/api/v1/rooms_controller.rb:8]: That will make it:
`
def create
  return ensure_authenticated
  # Rest of the code...
end
`

Comment (farhatahmad) [reply app/controllers/api/v1/rooms_controller.rb:8]: Nono let's drop the ensure authenticated completely from this. 

None of our other actions are secured yet, so it's better if we do them one shot

Comment (KH-Amir-TN) [reply app/controllers/api/v1/rooms_controller.rb:8]: Alright, I'll remove the check entirely.

File: spec/controllers/rooms_controller_spec.rb [hunk 5]
```diff
@@ -59,9 +59,29 @@
...
+      }
+    end
+
+    it 'creates a room for the authenticated user' do
+      session[:user_id] = user.id
+      expect { post :create, params: room_params }.to change { user.rooms.count }.from(0).to(1)
+      expect(response).to have_http_status(:created)
+    end
+
+    it 'returns :unauthorized for unauthenticated create requests' do
```
[the remark this answers is not in the document]
Comment (KH-Amir-TN) [reply spec/controllers/rooms_controller_spec.rb:81]: Yup, I'll remove it.
Thank you!

**Review by hadicheaito1 — APPROVED** [review]

**Review by farhatahmad — APPROVED** [review]
